File: day23.c
/* Just to compare speed */

#include <stdio.h>
#include <stdint.h>

#define BASE 1000001
#define NROUNDS 10000000

int mod(int a, int b)
{
    int r = a % b;
    return r < 0 ? r + b : r;
}

void pickup(int_fast32_t *cups, int_fast32_t current, int_fast32_t *pc) {
    pc[0] = cups[current];
    pc[1] = cups[pc[0]];
    pc[2] = cups[pc[1]];
    cups[current] = cups[pc[2]];
}

void insert(int_fast32_t *cups, int_fast32_t dest, int_fast32_t *pc) {
    cups[pc[2]] = cups[dest];
    cups[dest] = pc[0];
    cups[pc[0]] = pc[1];
    cups[pc[1]] = pc[2];
}

void pp(int_fast32_t *cups, int_fast32_t current) {
    printf("cups: (%d) ", current);
    int cup = cups[current];
    while(cup != current) {
        printf("%d ", cup);
        cup = cups[cup];
    }
    printf("\n");
}


int_fast32_t play(int_fast32_t *cups, int_fast32_t current, int_fast32_t *pc) {
    pickup(cups, current, pc);

    int_fast32_t dest = mod(current-1, BASE);

    while(dest == pc[0] || dest == pc[1] || dest == pc[2] || dest == 0)
        dest = mod(dest-1, BASE);
        
    insert(cups, dest, pc);

    return cups[current];
}

int_fast32_t main(int_fast32_t argc, char **argv) {
    int_fast32_t cups[BASE];
    int_fast32_t current = 3;
    static int_fast32_t input[] = {3,1,8,9,4,6,5,7,2};
    
    int_fast32_t pc[] = {0,0,0};

    /* Build the list of cups */
    for(int_fast32_t i=0; i<8; i++) {
        cups[input[i]] = input[i+1];
    }
    cups[2] = 10;
    
    for(int_fast32_t i=10; i<BASE-1; i++) {
        cups[i] = i+1;
    }
    cups[BASE-1] = 3;



    for(int_fast32_t i=0; i<NROUNDS; i++) {    
        current = play(cups, current, pc);
    }
    int_fast32_t c1 = cups[1];
    int_fast32_t c2 = cups[c1];

    printf("%d %d %llu\n", c1, c2, (unsigned long long)c1*c2);

    return 0;
}

File: day23.py
import array

def make_cups(lst):
    cups = array.array("Q", [0] * (len(lst)+1))
    for i in range(len(lst)-1):
        cups[lst[i]] = lst[i+1]
    cups[lst[-1]] = lst[0]
    return cups

def pp(cups, current):
    print(f"cups: ({current})", end=" ")

    cup = cups[current]

    while cup != current:
        print(cup, end=" ")
        cup = cups[cup]
    print()

def pickup(cups, current):
    pc = [0] * 3

    pc[0] = cups[current]
    pc[1] = cups[pc[0]]
    pc[2] = cups[pc[1]]
    cups[current] = cups[pc[2]]

    return pc

def insert(cups, dest, pc):
    tmp = cups[dest]
    cups[dest] = pc[0]
    cups[pc[0]] = pc[1]
    cups[pc[1]] = pc[2]
    cups[pc[2]] = tmp



def play(cups, current, base, display=True):
    if display:
        pp(cups, current)
    
    pickup_cups = pickup(cups, current)
    if display:
        print("pick up:", pickup_cups)
    
    dest = (current-1) % base
    while dest in pickup_cups or dest == 0:
        dest = (dest-1) % base
    if display:
        print("destination:", dest)

    insert(cups, dest, pickup_cups)

    return cups[current]

# PART 1
# cups = make_cups([int(x) for x in "318946572"])
# current = 3
# base = 10

# for i in range(100):
#     print("--", "Move", i+1, "--")
#     current = play(cups, current, base, display=True)
#     print()

# print("-- final --")
# pp(cups, current)

#PART 2
cups2 = make_cups([int(x) for x in "318946572"] + list(range(10,1000001)))
base2 = 1000001
current = 3
for i in range(10000000):
    current = play(cups2, current, base2, display=False)

c1 = cups2[1]
c2 = cups2[c1]
print(c1, c2, c1*c2)